### PR TITLE
Sw 271 implement staging environment

### PR DESCRIPTION
**Pull request by zaifastafa**

### CODE REVIEW

File: FinSearchUnified/Components/Environment.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    }
+    /**
+     * @return bool|null
+     * @throws Zend_Cache_Exception
+     */
+    private function getStagingFlagFromShopConfig()
+    {
+        /** @var ConfigLoader $configLoader */
+        $configLoader = Shopware()->Container()->get('fin_search_unified.config_loader');
+        return $configLoader->isStagingShop();
```
Comment (TheKeymaster) [inline FinSearchUnified/Components/Environment.php]: Add a space before the return statement.

File: FinSearchUnified/Tests/Helper/StaticHelperTest.php [hunk 2]
```diff
@@ -55,6 +54,125 @@ protected function tearDown()
...
+     *
+     * @param bool $isActive
+     * @param string $shopKey
+     * @param string|null $param
+     * @param bool $stagingFlag
+     * @param bool $configStaging
+     *
+     * @throws Zend_Cache_Exception
+     */
+    public function testisFindologicActive(
```
Comment (TheKeymaster) [inline FinSearchUnified/Tests/Helper/StaticHelperTest.php]: Rename to `testIsFindologicActive`.

File: FinSearchUnified/Tests/Helper/StaticHelperTest.php [hunk 3]
```diff
@@ -55,6 +54,125 @@ protected function tearDown()
...
+    /**
+     * Data provider for checking isFindologicActive behavior
+     *
+     * @return array
+     */
+    public function isFindologicActiveDataprovider()
+    {
+        return [
+            'Plugin is deactivated' => [
+                'ActivateFindologic' => false,
```
Comment (TheKeymaster) [inline FinSearchUnified/Tests/Helper/StaticHelperTest.php]: camelCase your data provider keys and make sure they have the same name as the method params that use these data providers.

File: FinSearchUnified/Tests/Helper/StaticHelperTest.php [hunk 4]
```diff
@@ -55,6 +54,125 @@ protected function tearDown()
...
+                'ShopKey' => '0000000000000000ZZZZZZZZZZZZZZZZ',
```
Comment (TheKeymaster) [inline FinSearchUnified/Tests/Helper/StaticHelperTest.php]: Please use some theoretically real shopkey (e.g. `80AB18D4BE2654E78244106AD315DC2C`)

File: FinSearchUnified/Tests/Helper/StaticHelperTest.php [hunk 5]
```diff
@@ -55,6 +54,125 @@ protected function tearDown()
...
+            ],
+            'FINDOLOGIC is active and shopkey is empty' => [
+                'ActivateFindologic' => true,
+                'ShopKey' => '',
+                'QueryParameter' => null,
+                'Staging' => false,
+                'ConfigStaging' => true,
+
+            ],
+            'FINDOLOGIC is active without query parameter' => [
```
Comment (TheKeymaster) [inline FinSearchUnified/Tests/Helper/StaticHelperTest.php]: Clarify that this case checks that FINDOLOGIC will not be active if it is a staging shop with this configuration.

File: FinSearchUnified/Tests/Helper/StaticHelperTest.php [hunk 6]
```diff
@@ -55,6 +54,125 @@ protected function tearDown()
...
+    public function isFindologicActiveDataprovider()
+    {
+        return [
+            'Plugin is deactivated' => [
+                'ActivateFindologic' => false,
+                'ShopKey' => '0000000000000000ZZZZZZZZZZZZZZZZ',
+                'QueryParameter' => null,
+                'Staging' => false,
+                'ConfigStaging' => true,
+
```
Comment (TheKeymaster) [inline FinSearchUnified/Tests/Helper/StaticHelperTest.php]: Remove these empty lines (in each case).

File: FinSearchUnified/Tests/Helper/StaticHelperTest.php [hunk 7]
```diff
@@ -55,6 +54,125 @@ protected function tearDown()
...
+            ],
+            'FINDOLOGIC is active with query parameter' => [
+                'ActivateFindologic' => true,
+                'ShopKey' => '0000000000000000ZZZZZZZZZZZZZZZZ',
+                'QueryParameter' => 'on',
+                'Staging' => true,
+                'ConfigStaging' => false,
+
```
Comment (TheKeymaster) [inline FinSearchUnified/Tests/Helper/StaticHelperTest.php]: I am missing some cases such as:
* Plugin inactive, shopkey set, not staging mode, no query param.  

Or successful cases even:
* Plugin active, shopkey set, not staging mode, no query param.

File: FinSearchUnified/Tests/Helper/StaticHelperTest.php [hunk 8]
```diff
@@ -119,14 +237,14 @@ public static function configDataProvider()
...
-            'FINDOLOGIC is active in search' => [
+            'FINDOLOGIC is active on search page' => [
                 'ActivateFindologic' => true,
                 'ShopKey' => '0000000000000000ZZZZZZZZZZZZZZZZ',
                 'ActivateFindologicForCategoryPages' => false,
                 'findologicDI' => false,
                 'isSearchPage' => true,
                 'isCategoryPage' => false,
-                'expected' => false
+                'expected' => true
```
Comment (TheKeymaster) [inline FinSearchUnified/Tests/Helper/StaticHelperTest.php]: This case wasn't fixed, but rather the condition was changed. Make sure this still returns `false`. You may need some more setup so the `Environment` returns what you need.

File: FinSearchUnified/Tests/Helper/StaticHelperTest.php [hunk 9]
```diff
@@ -135,7 +253,16 @@ public static function configDataProvider()
                 'findologicDI' => false,
                 'isSearchPage' => false,
                 'isCategoryPage' => true,
-                'expected' => false
+                'expected' => true
```
Comment (TheKeymaster) [inline FinSearchUnified/Tests/Helper/StaticHelperTest.php]: This case wasn't fixed, but rather the condition was changed. Make sure this still returns `false`. You may need some more setup so the `Environment` returns what you need.

File: FinSearchUnified/Tests/Helper/StaticHelperTest.php [hunk 10]
```diff
@@ -135,7 +253,16 @@ public static function configDataProvider()
...
+            ],
+            'FINDOLOGIC is active on category pages but it was disabled in the plugin config' => [
```
Comment (TheKeymaster) [inline FinSearchUnified/Tests/Helper/StaticHelperTest.php]: Why was this case added when it basically does the same as the case above?

File: FinSearchUnified/Tests/Helper/StaticHelperTest.php [hunk 11]
```diff
@@ -388,27 +513,6 @@ public function testUseShopSearchForBackendRequests()
         // Assign mocked session variable to application container
         Shopware()->Container()->set('front', $front);
 
-        // Create Mock object for Shopware Config
```
Comment (TheKeymaster) [inline FinSearchUnified/Tests/Helper/StaticHelperTest.php]: Why was this code block removed? Maybe this code now runs because it fails somewhere else and does not test what it should actually test?

File: FinSearchUnified/Tests/Helper/StaticHelperTest.php [hunk 12]
```diff
@@ -678,7 +782,6 @@ public function testCreateRangeSliderFacetMethod(
                 $filter->addChild($key, $value);
             }
         }
-
```
Comment (TheKeymaster) [inline FinSearchUnified/Tests/Helper/StaticHelperTest.php]: Re-add this line because it separates blocks.

File: FinSearchUnified/Tests/Helper/StaticHelperTest.php [hunk 13]
```diff
@@ -300,44 +424,37 @@ public function testUseShopSearch(
...
 
             // Create Mock object for Shopware Session
-            $session = $this->getMockBuilder(Session::class)
-                ->setMethods(['offsetGet'])
-                ->getMock();
+            $session = $this->createMock(Session::class);
             $session->expects($this->atLeastOnce())
                 ->method('offsetGet')
                 ->willReturnMap($sessionArray);
-
```
Comment (TheKeymaster) [inline FinSearchUnified/Tests/Helper/StaticHelperTest.php]: Re-add this newline as it separates the blocks.

File: FinSearchUnified/Tests/Helper/StaticHelperTest.php [hunk 14]
```diff
@@ -276,17 +405,12 @@ public function testUseShopSearch(
             ['ActivateFindologicForCategoryPages', $isActiveForCategory],
             ['IntegrationType', $checkIntegration ? Constants::INTEGRATION_TYPE_DI : Constants::INTEGRATION_TYPE_API]
         ];
-
         $request = new RequestHttp();
```
Comment (TheKeymaster) [inline FinSearchUnified/Tests/Helper/StaticHelperTest.php]: Test case `If no request is given => should return true` has not been added, as the request seems to always be created.

File: FinSearchUnified/Tests/Helper/StaticHelperTest.php [hunk 15]
```diff
@@ -55,6 +54,125 @@ protected function tearDown()
         Shopware()->Container()->load('session');
         Shopware()->Container()->reset('config');
         Shopware()->Container()->load('config');
+        Shopware()->Container()->reset('fin_search_unified.config_loader');
+        Shopware()->Container()->load('fin_search_unified.config_loader');
+    }
+
+    /**
+     * Data provider for checking isFindologicActive behavior
```
Comment (TheKeymaster) [inline FinSearchUnified/Tests/Helper/StaticHelperTest.php]: This docblock comment does not add any value.

File: FinSearchUnified/Tests/Helper/StaticHelperTest.php [hunk 16]
```diff
@@ -135,7 +253,16 @@ public static function configDataProvider()
...
+            ],
+            'FINDOLOGIC is active on category pages but it was disabled in the plugin config' => [
+                'ActivateFindologic' => true,
+                'ShopKey' => '0000000000000000ZZZZZZZZZZZZZZZZ',
+                'ActivateFindologicForCategoryPages' => true,
+                'findologicDI' => false,
+                'isSearchPage' => false,
+                'isCategoryPage' => true,
+                'expected' => true
             ]
```
Comment (TheKeymaster) [inline FinSearchUnified/Tests/Helper/StaticHelperTest.php]: I am missing the test case `In any other case where none of these apply => should return false`.

**Comment by TheKeymaster** [comment]
@rababfatemah please also make sure to fix the merge conflicts.

File: FinSearchUnified/Tests/Helper/StaticHelperTest.php (hunk 11 above)
Comment (zaifastafa) [reply FinSearchUnified/Tests/Helper/StaticHelperTest.php]: It was removed because it did not have any significant impact on the test outcome, since we are testing for the backend request only in this case.

File: FinSearchUnified/Tests/Helper/StaticHelperTest.php (hunk 14 above)
Comment (zaifastafa) [reply FinSearchUnified/Tests/Helper/StaticHelperTest.php]: there is already a separate test for this case called `testUseShopSearchWhenRequestIsNull`

File: FinSearchUnified/Tests/Helper/StaticHelperTest.php (hunk 16 above)
Comment (zaifastafa) [reply FinSearchUnified/Tests/Helper/StaticHelperTest.php]: This case is also already covered in one of the cases above.

File: FinSearchUnified/Components/ConfigLoader.php [hunk 17]
```diff
@@ -71,7 +70,7 @@ private function isSuccessful(Response $response)
     {
         $restype = (string)$response->getStatusCode();
 
-        return $restype[0] === '2' || $restype[0] === '1';
+        return strpos($restype, '2') === 0 || strpos($restype, '1') === 0;
```
Comment (TheKeymaster) [inline FinSearchUnified/Components/ConfigLoader.php]: If we are already refactoring this code, we could also cast the status code to an integer and ensure that it is bigger or equal 100 and lower than 300.
Please add unit-tests for code that you have changed to ensure your change does still work as expected.

Comment (zaifastafa) [reply FinSearchUnified/Components/ConfigLoader.php]: The logic is still the same for this change, its just a bit of optimization in terms of readability

Comment (TheKeymaster) [reply FinSearchUnified/Components/ConfigLoader.php]: I just checked the `ConfigLoaderTest` and it seems like this method is already covered in tests. For future changes please keep in mind that even if it is a small change or even just a change that makes the code more readable, there should always be an unit-test associated to it. Because if something goes wrong (for whatever reason), the build will break and will notify us.

Comment (zaifastafa) [reply FinSearchUnified/Components/ConfigLoader.php]: Sure will keep that in mind for future. Although I knew that the case was already covered as I wrote that myself. I was about to tell you about it but got occupied in some other task
